Orleans.Consensus/Log/StreamLog.cs

```diff
@@ -0,0 +1,191 @@
+﻿namespace Orleans.Consensus.Log
+{
+    using Orleans.Consensus.Contract.Log;
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Threading.Tasks;
+
+    public class StreamLog<TOperation> : IPersistentLog<TOperation>, IDisposable
+    {
+        const int CACHE_SIZE = 1000;
+
+        Stream stream;
+        ISerializer<LogEntry<TOperation>> serializer;
+        BTree<long, long> bTreeIndex;
+        IDictionary<long, LogEntry<TOperation>> cache;
+
+        public StreamLog(Stream stream, ISerializer<LogEntry<TOperation>> serializer)
+        {
+            if (null == stream) throw new ArgumentNullException(nameof(stream));
+            if (null == serializer) throw new ArgumentNullException(nameof(serializer));
+
+            this.stream = stream;
+            this.serializer = serializer;
+            this.bTreeIndex = new BTree<long, long>();
+            this.cache = new Dictionary<long, LogEntry<TOperation>>(CACHE_SIZE);
+            BuildIndex();
+        }
+
+        void BuildIndex()
+        {
+            stream.Position = 0;
+            var first = true;
+            while (stream.Position < stream.Length)
+            {
+                var position = stream.Position;
+                var entry = serializer.Deserialize(stream).Result;
```

Comment:
the original motivation for making the serializer async was to maintain async all the way down. Other serializers might support async writing to the stream, alternatively the serializer could write to a memory stream, which could then write to the physical steam async (as you suggested).

In some recent performance testing I found that async = slow.